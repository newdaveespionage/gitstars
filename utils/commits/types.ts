export type Commit = {
  id: string
  name: string
  date: string
  message: string
}
