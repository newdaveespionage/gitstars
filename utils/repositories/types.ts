export type Repository = {
  id?: string
  name: string
  owner: string
  url: string
  stars: number
  description: string
}
