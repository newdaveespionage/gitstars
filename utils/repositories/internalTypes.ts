export type RawRepository = {
  id: string
  name: string
  owner: {
    login: string
  }
  html_url: string
  stargazers_count: number
  description: string
}
