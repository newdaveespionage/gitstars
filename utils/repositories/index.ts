import { RawRepository } from './internalTypes'
import { Repository } from './types'

export const standardizeRepositories = (
  rawRepos: RawRepository[]
): Repository[] =>
  rawRepos.map((rawRepo) => ({
    id: rawRepo.id,
    name: rawRepo.name,
    owner: rawRepo.owner.login,
    url: rawRepo.html_url,
    stars: rawRepo.stargazers_count,
    description: rawRepo.description,
  }))
