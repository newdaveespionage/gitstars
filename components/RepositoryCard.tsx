import React from 'react'
import { colors } from '../theme'
import { Repository } from '../utils/repositories/types'
import Commits from './Commits'
import Image from 'next/image'

const RespositoryCard = ({
  name,
  owner,
  url,
  stars,
  description,
}: Repository): JSX.Element => (
  <div className="card" title={`${name}: ${description}`}>
    <h3 className="title">{name}</h3>
    <p className="description">{description}</p>
    <div className="details">
      <p>Owned by {owner}</p>
      <a href={url} target="_blank" rel="noopener noreferrer">
        {url}
      </a>
      <div className="star-container">
        <Image
          src="/icons/star.svg"
          height={18}
          width={20}
          alt="star icon"
          className="star-icon"
        />
        <span className="star-count" title="total stars">
          {stars}
        </span>
      </div>
    </div>
    <Commits owner={owner} name={name} />
    <style jsx>{`
      .card {
        flex-basis: 30%;
        margin: 1rem;
        padding: 1rem;
        text-align: left;
        color: inherit;
        border: 1px solid ${colors.border};
        transition: color 0.15s ease, border-color 0.15s ease;
        align-self: flex-start;
      }

      .card:hover,
      .card:focus,
      .card:active {
        color: ${colors.default};
        border-color: ${colors.blue100};
      }

      .details {
        margin-top: 1.5rem;
      }

      .card h3 {
        margin: 0 0 1rem 0;
        font-size: 1.5rem;
      }

      .card a {
        color: ${colors.blue600};
        text-decoration: none;
      }

      .card a:hover,
      .card a:focus,
      .card a:active {
        text-decoration: underline;
      }

      .title {
        margin: 0;
        line-height: 1.15;
        font-size: 4rem;
      }

      .title,
      .description {
        text-align: center;
      }

      .description {
        line-height: 1.5;
        font-size: 1.2rem;
        height: 2rem;
        max-width: 400px;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
        margin: 1rem 0;
      }

      .star-container {
        margin: 1rem 0;
      }

      .star-icon {
        vertical-align: middle;
      }

      .star-count {
        font-size: 1.2rem;
      }
    `}</style>
  </div>
)

export default RespositoryCard
