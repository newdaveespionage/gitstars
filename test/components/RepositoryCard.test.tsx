import React from 'react'
import { render } from '../testUtils'
import RepositoryCard from '../../components/RepositoryCard'
import { Repository } from '../../utils/repositories/types'

export const repositoryCardTestProps: Repository = {
  owner: 'Test Owner',
  name: 'Test Name',
  url: 'https://www.github.com/testowner/testrepo',
  stars: 10000000000,
  description: 'ten billion stars repo',
}

describe('CommitItem', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      <RepositoryCard {...repositoryCardTestProps} />,
      {}
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
