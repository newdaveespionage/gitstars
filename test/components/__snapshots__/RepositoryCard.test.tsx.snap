// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommitItem matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-3205085322 card"
    title="Test Name: ten billion stars repo"
  >
    <h3
      class="jsx-3205085322 title"
    >
      Test Name
    </h3>
    <p
      class="jsx-3205085322 description"
    >
      ten billion stars repo
    </p>
    <div
      class="jsx-3205085322 details"
    >
      <p
        class="jsx-3205085322"
      >
        Owned by Test Owner
      </p>
      <a
        class="jsx-3205085322"
        href="https://www.github.com/testowner/testrepo"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://www.github.com/testowner/testrepo
      </a>
      <div
        class="jsx-3205085322 star-container"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjAiIGhlaWdodD0iMTgiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <noscript />
          <img
            alt="star icon"
            class="star-icon"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
        <span
          class="jsx-3205085322 star-count"
          title="total stars"
        >
          10000000000
        </span>
      </div>
    </div>
    <div
      class="jsx-3680457408"
    >
      <button
        class="jsx-3680457408 load-cta"
        title="use this button to load commits from the past 24 hours"
      >
        Tap to load commits
      </button>
    </div>
  </div>
</DocumentFragment>
`;
